### PR TITLE
ensure that the BOM is used to sniff the encoding first

### PR DESCRIPTION
**Pull request by BurnzZ**
Resolves https://github.com/scrapinghub/web-poet/issues/64.

Based from the https://github.com/scrapy/scrapy/issues/5601 Scrapy fix.

### CODE REVIEW

**Review by Gallaecio — APPROVED** [review]

File: web_poet/page_inputs/http.py [hunk 1]
```diff
@@ -261,6 +263,10 @@ def urljoin(self, url: Union[str, _RequestUrl, _ResponseUrl]) -> _RequestUrl:
         *self*."""
         return _RequestUrl(urljoin(self._base_url, str(url)))
 
+    @memoizemethod_noargs
+    def _bom_encoding(self):
+        return read_bom(self.body)[0]
```
Comment (kmike) [inline web_poet/page_inputs/http.py:268]: I wonder if this code should be moved to HttpResponseBody class

Comment (BurnzZ) [reply web_poet/page_inputs/http.py]: Good call, moved it in https://github.com/scrapinghub/web-poet/pull/81/commits/c39d176050229d4ccb3bbf639ad252e5943fa23e.

File: web_poet/page_inputs/http.py [hunk 2]
```diff
@@ -36,12 +37,16 @@ class HttpRequestBody(bytes):
...
+        """Returns the encoding from the byte order mark if present."""
+        return read_bom(self)[0]
+
     def declared_encoding(self) -> Optional[str]:
         """Return the encoding specified in meta tags in the html body,
         or ``None`` if no suitable encoding was found"""
         return html_body_declared_encoding(self)
 
-    def json(self):
+    def json(self) -> Optional[str]:
```
Comment (kmike) [inline web_poet/page_inputs/http.py:49]: This type annotation is wrong, any object can be returned, as it's result of json.loads

File: web_poet/page_inputs/http.py [hunk 3]
```diff
@@ -233,17 +238,18 @@ def _selector_input(self) -> str:
...
             self._encoding
+            or self._body_bom_encoding()
             or self._headers_declared_encoding()
             or self._body_declared_encoding()
             or self._body_inferred_encoding()
         )
 
     @memoizemethod_noargs
-    def json(self):
+    def json(self) -> Optional[str]:
```
Comment (kmike) [inline web_poet/page_inputs/http.py:252]: the same, it can return anything :)

File: web_poet/page_inputs/http.py (hunk 2 above)
Comment (BurnzZ) [reply web_poet/page_inputs/http.py]: Nice catch! I was on a copy-pasting spree on the past commit. 🤦 

Although I'm curious as to why mypy wasn't able to detect this one as the tests were comparing it to a `dict`.

File: web_poet/page_inputs/http.py [hunk 4]
```diff
@@ -46,7 +46,7 @@ def declared_encoding(self) -> Optional[str]:
         or ``None`` if no suitable encoding was found"""
         return html_body_declared_encoding(self)
 
-    def json(self) -> Optional[str]:
+    def json(self) -> Optional[Any]:
```
Comment (kmike) [inline web_poet/page_inputs/http.py:49]: Optional[Any] is the same as Any :)

**Comment by kmike** [comment]
Thanks @BurnzZ!

**Pull request merged**